lib/ssl/src/tls_handshake_1_3.erl

```diff
@@ -3051,15 +3059,27 @@ select_client_cert_key_pair(Session0, [#{private_key := Key, certs := [Cert| _]
                                              own_certificates = EncodedChain,
                                              private_key = Key
                                             };
-                        {error, _, not_in_auth_domain} ->
+                        {error, EncodedChain, not_in_auth_domain} ->
+                            Plausible = plausible_missing_chain(EncodedChain, Plausible0,
+                                                                SelectedSignAlg, Key, Session0),
                             select_client_cert_key_pair(Session0, Rest, ServerSignAlgs, ServerSignAlgsCert,
-                                                        ClientSignAlgs, CertDbHandle, CertDbRef, CertAuths)
+                                                        ClientSignAlgs, CertDbHandle, CertDbRef, CertAuths,
+                                                        Plausible)
                     end;
                 _ ->
                     select_client_cert_key_pair(Session0, Rest, ServerSignAlgs, ServerSignAlgsCert, ClientSignAlgs,
-                                                CertDbHandle, CertDbRef, CertAuths)
+                                                CertDbHandle, CertDbRef, CertAuths, Plausible0)
             end;
         {error, _} ->
             select_client_cert_key_pair(Session0, Rest, ServerSignAlgsCert, ServerSignAlgsCert, ClientSignAlgs,
-                                        CertDbHandle, CertDbRef, CertAuths)
+                                        CertDbHandle, CertDbRef, CertAuths, Plausible0)
     end.
+
+plausible_missing_chain([_] = EncodedChain, undefined, SignAlg, Key, Session0) ->
```

Comment:
```suggestion
plausible_missing_chain([_ | _] = EncodedChain, undefined, SignAlg, Key, Session0) ->
```
Does this need to be restricted to only one single cert in the EncodedChain?

Reply:
Well, the assumption I made was that the case where the domain adherence check becomes a problem is when the client is "misconfigured" to only supply its own cert and not the other certs in the chain.  As in this case, we cannot perform the check,  but the server may be able to verify it anyway (TLS-1.3 spec mandates to try and do that). So in this case we don't really know and it is worth sending what we got. I could send a default always if the check fails, but that kind of makes the check pointless, and the TLS-1.3 spec also says the client SHOULD adhere to the server's auth domain.

Reply:
@jjcarstens do you have an actual use case when only part of the chain is supplied? (Root cert can safely be left out).

Reply:
Maybe. We roughly follow the same setup for [AWS IOT Client authentication using X.509 Certificates](https://docs.aws.amazon.com/iot/latest/developerguide/x509-client-certs.html). The server authentication to complete handshake is the same - The device has the certificate chain and includes it in `cacerts` of the request. But the device/client also includes the signer CA used to sign it's device certificate as well which is typically a self-signed and managed CA and also put into the `cacerts` request.

This signer CA is not required by our server and is only used when the device certificate being presented is unknown (i.e. not pinned in our DB) and is required to be registered with our service first (also like [AWS IOT requires registering CA's first](https://docs.aws.amazon.com/iot/latest/developerguide/register-CA-cert.html)). In that situation, we'll run PKIX validation with this signer CA and presented client cert before allowing the connection on the server. If we already know about the client certificate (after looking up its calculated pinned value in the DB), then the signer CA is simply ignored.

It's this extra, unchained signer CA that is getting us with the TLS 1.3 changes. I also don't claim to be an SSL/TLS expert, so it could be that I'm doing things completely wrong and don't fully understand the standards.

Reply:
Well,  TLS-1.3  RFC says  this about the extension:

  ```
 The client MAY send the "certificate_authorities" extension in the
   ClientHello message.  The server MAY send it in the
   CertificateRequest message.
```

```
4.4.2.3.  Client Certificate Selection

   The following rules apply to certificates sent by the client:

   -  The certificate type MUST be X.509v3 [RFC5280], unless explicitly
      negotiated otherwise (e.g., [RFC7250]).

   -  If the "certificate_authorities" extension in the
      CertificateRequest message was present, at least one of the
      certificates in the certificate chain SHOULD be issued by one of
      the listed CAs.

   -  The certificates MUST be signed using an acceptable signature
      algorithm, as described in Section 4.3.2.  Note that this relaxes
      the constraints on certificate-signing algorithms found in prior
      versions of TLS.

   -  If the CertificateRequest message contained a non-empty
      "oid_filters" extension, the end-entity certificate MUST match the
      extension OIDs that are recognized by the client, as described in
      Section 4.2.5.

```

So I think that your use case fits better with having an option to disable the sending of the certificat_authorties extension.

In TLS 1.2 this particular extension is not present but an equivalent value exists sent in the certificate_reques, although the RFC is less clear on how to act, and before it was just ignored, so we let it be backward compatible.